SimPEG/EM/Analytics/NSEM.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+from scipy.constants import epsilon_0
+from scipy.constants import mu_0
+from SimPEG.EM.Utils.EMUtils import k
+from SimPEG.EM.Utils.EMUtils import omega
+
+__all__ = [
+    'MT_LayeredEarth'
+]
+
+
+# Evaluate Impedance Z of a layer
+_ImpZ = lambda f, mu, k: omega(f)*mu/k
+
+# Complex Cole-Cole Conductivity - EM utils
+_PCC = lambda siginf, m, t, c, f: siginf*(1.-(m/(1.+(1j*omega(f)*t)**c)))
+
+# matrix P relating Up and Down components with E and H fields
+_P = lambda z: np.matrix([[1., 1, ], [-1./z, 1./z]], dtype='complex_')
+_Pinv = lambda z: np.matrix([[1., -z], [1., z]], dtype='complex_')/2.
+
+# matrix T for transition of Up and Down components accross a layer
+_T = lambda h, k: np.matrix([[np.exp(1j*k*h), 0.], [0., np.exp(-1j*k*h)]], dtype='complex_')
+_Tinv = lambda h, k: np.matrix([[np.exp(-1j*k*h), 0.], [0., np.exp(1j*k*h)]], dtype='complex_')
+
+# Propagate Up and Down component for a certain frequency & evaluate E and H field
+def _Propagate(f, thickness, sig, chg, taux, c, mu_r, eps_r, n):
+
+    if isinstance(sig, float):
+        sigmodel = np.r_[sig]
+    else:
+        sigmodel = sig
+
+    if isinstance(eps_r, float):
+        epsmodel = np.ones_like(sigmodel)*eps_r
+    else:
+        epsmodel = eps_r
+
+    if isinstance(mu_r, float):
+        mumodel = np.ones_like(sigmodel)*mu_r
+    else:
+        epsmodel = mu_r
+
+    sigcm = np.zeros_like(sigmodel, dtype='complex_')
+    if chg == 0. or taux == 0. or c == 0.:
+        sigcm = sigmodel
+    else:
+        for j in range(1, len(sigcm)):
+            sigcm[j] = _PCC(sigmodel[j], chg[j], taux[j], c[j], f)
+
+    sigcm = np.append(np.r_[0.], sigcm)
+    mu = np.append(np.r_[1.], mumodel)*mu_0
+    eps = np.append(np.r_[1.], epsmodel)*epsilon_0
+    H = np.append(np.r_[1.2*(1e5)], thickness)
+
+    K = k(f, sigcm, mu, eps)
+    Z = _ImpZ(f, mu, K)
+
+    EH = np.matrix(np.zeros((2, n+1), dtype='complex_'), dtype='complex_')
+    UD = np.matrix(np.zeros((2, n+1), dtype='complex_'), dtype='complex_')
+
+    UD[1, -1] = 1.
+
+    for i in range(-2, -(n+2), -1):
+
+        UD[:, i] = _Tinv(H[i+1], K[i])*_Pinv(Z[i])*_P(Z[i+1])*UD[:, i+1]
+        UD = UD/((np.abs(UD[0, :]+UD[1, :])).max())
+
+    for j in range(0, n+1):
+        EH[:, j] = np.matrix([[1., 1, ], [-1./Z[j], 1./Z[j]]])*UD[:, j]
+
+    return UD, EH, Z, K
+
+# Utils to compute the apparent impedance over a layered Earth Model
+def MT_LayeredEarth(freq, thickness, sig, nlayer, return_type='Res-Phase', chg=0., tau=0., c=0., mu_r=1., eps_r=1.):
```

Comment:
my suggestion here would be to remove `nlayer` and add a check that sigma and thickness are the correct lengths 
```python
assert len(sigma) == len(thickness) + 1, (
    "len(sigma) should be 1 longer than len(thickness), "
    "the vectors provided  are len(sigma): {}, len(thickness): {}".format(len(sigma), len(thickness))
)
```